src/main/java/net/neoforged/neoforge/transfer/transaction/Transaction.java

```diff
@@ -0,0 +1,154 @@
+ * <pre>{@code
+ * try (Transaction outerTransaction = Transaction.openOuter()) {
+ *     // (A) some transaction operations
+ *     try (Transaction nestedTransaction = outerTransaction.openNested()) {
+ *         // (B) more operations
+ *         nestedTransaction.commit(); // Validate the changes that happened in this transaction.
+ *                                     // This is a nested transaction, so changes will only be applied if the outer
+ *                                     // transaction is committed too.
+ *     }
+ *     // (C) even more operations
+ *     outerTransaction.commit(); // This is an outer transaction: changes (A), (B) and (C) are applied.
+ * }
+ * // If we hadn't committed the outerTransaction, all changes (A), (B) and (C) would have been reverted.
+ * }</pre>
+ *
+ * <p>Participants are responsible for upholding this contract themselves, by using {@link #addCloseCallback}
+ * to react to transaction close events and properly validate or revert changes.
+ * Any action that modifies state outside of the transaction, such as calls to {@code markDirty()} or neighbor updates,
+ * should be deferred until {@linkplain #addOuterCloseCallback after the outer transaction is closed}
+ * to give every participant a chance to react to transaction close events.
+ *
+ * <p>This is very low-level for most applications, and most participants should subclass {@link SnapshotParticipant}
+ * that will take care of properly maintaining their state.
+ *
+ * <p>Participants should generally be passed a {@link TransactionContext} parameter instead of the full {@code Transaction},
+ * to make sure they don't call {@link #abort}, {@link #commit} or {@link #close} mistakenly.
+ *
+ * <p>Every transaction is only valid on the thread it was opened on,
+ * and attempts to use it on another thread will throw an exception.
+ * Consequently, transactions can be concurrent across multiple threads, as long as they don't share any state.
+ */
+@ApiStatus.NonExtendable
+public interface Transaction extends AutoCloseable, TransactionContext {
+    /**
+     * Open a new outer transaction.
+     *
+     * @throws IllegalStateException If a transaction is already active on the current thread.
+     */
+    static Transaction openOuter() {
+        return TransactionManagerImpl.MANAGERS.get().openOuter();
+    }
+
+    /**
+     * @return True if a transaction is open or closing on the current thread, and false otherwise.
+     */
+    static boolean isOpen() {
+        return getLifecycle() != Lifecycle.NONE;
+    }
+
+    /**
+     * @return The current lifecycle of the transaction stack on this thread.
+     */
+    static Lifecycle getLifecycle() {
+        return TransactionManagerImpl.MANAGERS.get().getLifecycle();
+    }
+
+    /**
+     * Open a nested transaction if {@code maybeParent} is non-null, or an outer transaction if {@code maybeParent} is null.
+     */
+    static Transaction openNested(@Nullable TransactionContext maybeParent) {
```

Comment:
I think initial adoption could be made easier by merging both `openOuter` and `openNested` into a singular method:
```java
static Transaction open(@Nullable TransactionContext parent)
```
And then instead of having the weird "maybe parent" stuff with `null` redirecting to just being `openOuter`, we just have it so you explicitly pass null instead of adding the text "Outer" to the method. It also would have the benefit of making any code where people would be "abusing" `openNested` to do that already clearer about what is going on when it is the first level and really just an outer one.

Reply:
So the interesting history here is that the static `openNested` was added later than the original `openOuter` and `TransactionContext#openNested`. I agree with you though that we could just pass null. Would you then remove `TransactionContext#openNested` entirely and have everything go through `Transaction.open(@Nullable parent)`? I think that would work well.

Reply:
Passing `null` as the parent is already the standard for many helper methods that can work both as part of a transaction as well as standalone.

Reply:
Done.